tools/harness-simulation/posix/sniffer_sim/sniffer.py

```diff
@@ -28,104 +28,172 @@
 #
 
 import argparse
+from concurrent import futures
+import enum
+import grpc
+import ipaddress
 import logging
 import signal
-import time
 import pcap_codec
-import sys
 import threading
 
+from proto import sniffer_pb2
+from proto import sniffer_pb2_grpc
 import sniffer_transport
 
 
-class Sniffer:
-    """ Class representing the Sniffing node, whose main task is listening.
-    """
+class SnifferServicer(sniffer_pb2_grpc.Sniffer):
+    """ Class representing the Sniffing node, whose main task is listening. """
 
-    logger = logging.getLogger('sniffer.Sniffer')
+    logger = logging.getLogger('sniffer.SnifferServicer')
 
     RECV_BUFFER_SIZE = 4096
+    MAX_NODES_NUM = 33
 
-    def __init__(self, filename, channel):
-        self._pcap = pcap_codec.PcapCodec(filename, channel)
-
-        # Create transport
-        transport_factory = sniffer_transport.SnifferTransportFactory()
-        self._transport = transport_factory.create_transport()
+    class State(enum.Enum):
+        STOPPED = 0
+        RUNNING = 1
 
+    def _clear(self):
+        self._state = SnifferServicer.State.STOPPED
+        self._pcap = None
+        self._allowed_nodeids = None
+        self._transport = None
         self._thread = None
-        self._thread_alive = threading.Event()
         self._thread_alive.clear()
 
+    def __init__(self):
+        self._thread_alive = threading.Event()
+        self._mutex = threading.Lock()  # for self._allowed_nodeids
+        self._clear()
+
     def _sniffer_main_loop(self):
         """ Sniffer main loop. """
 
         self.logger.debug('Sniffer started.')
 
         while self._thread_alive.is_set():
+            # Avoid being blocked endlessly when there is no data
+            if not self._transport.ready(0.1):
+                continue
             data, nodeid = self._transport.recv(self.RECV_BUFFER_SIZE)
-            self._pcap.append(data)
+
+            self._mutex.acquire()
+            allowed_nodeids = self._allowed_nodeids
+            self._mutex.release()
+
+            # Equivalent to RF enclosure
+            if nodeid in allowed_nodeids:
+                self._pcap.append(data)
 
         self.logger.debug('Sniffer stopped.')
 
-    def start(self):
+    def Start(self, request, context):
         """ Start sniffing. """
 
+        self.logger.debug('call Start')
+
+        # Validate and change the state
+        if self._state != SnifferServicer.State.STOPPED:
+            return sniffer_pb2.StatusReply(status=sniffer_pb2.OPERATION_ERROR)
+        self._state = SnifferServicer.State.RUNNING
+
+        self._pcap = pcap_codec.PcapCodec(request.channel)
+
+        # Sniffer all nodes in default, i.e. there is no RF enclosure
+        self._allowed_nodeids = set(range(1, self.MAX_NODES_NUM + 1))
```

Comment:
Why not use `self._allowed_nodeids = None` as not filtering any node?

Reply:
Accepted for now.

But currently `self._allowed_nodeids` cannot be set to `None` again without restarting the sniffer. It is not a big problem but `set(range(1, self.MAX_NODES_NUM + 1))` has to be used again in `FilterNodes()`. Maybe change to `denied_nodeids` later? Use `set()` rather than `None` to represent the case without RF enclosure for consistency.

Reply:
Either way works for me.